src/router.ts

```diff
@@ -19,13 +19,21 @@ import {
 } from "./registry/registry";
 import { RegistryHTTPClient } from "./registry/http";
 
+function checkPermission(req: Request, capability: "push" | "pull"): void {
+  if (!req.credentials?.payload?.capabilities.includes(capability)) {
+    throw new ForbiddenError(req);
+  }
+}
+
 const v2Router = Router({ base: "/v2/" });
 
 v2Router.get("/", async (_req, _env: Env) => {
   return new Response();
 });
 
 v2Router.get("/_catalog", async (req, env: Env) => {
+  checkPermission(req, "pull");
```

Comment:
Could we move the checkPermission() calls to a more generic layer, perhaps in the authentication middleware? We could limit `pull` to request.method in `["GET", "HEAD"]` and `push` for the others?

Reply:
It turned out that the checking of capabilities was in the code for jwt tokens already all the time (`RegistryTokens.verifyPayload`). Thanks for your hint with the authentication middleware 🙏